File: laba3/task2/Golang/hasha_test.go
package main

import (
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/suite"
)

type HashASuite struct {
	suite.Suite
	hashTable *HashA
}

func (suite *HashASuite) SetupTest() {
	suite.hashTable = NewHashA()
}

func TestHashASuite(t *testing.T) {
	suite.Run(t, new(HashASuite))
}

func (suite *HashASuite) TestAddElement() {
	suite.hashTable.addElementA("key1", 100)
	assert.Equal(suite.T(), 1, suite.hashTable.getSize())

	suite.hashTable.addElementA("key2", 200)
	assert.Equal(suite.T(), 2, suite.hashTable.getSize())
}

func (suite *HashASuite) TestAddElementUpdate() {
	suite.hashTable.addElementA("key1", 100)
	suite.hashTable.addElementA("key1", 200)

	assert.Equal(suite.T(), 1, suite.hashTable.getSize())
	assert.Equal(suite.T(), 200, suite.hashTable.findElementA("key1"))
}

func (suite *HashASuite) TestFindExistingElement() {
	suite.hashTable.addElementA("test1", 42)
	suite.hashTable.addElementA("test2", 84)

	assert.Equal(suite.T(), 42, suite.hashTable.findElementA("test1"))
	assert.Equal(suite.T(), 84, suite.hashTable.findElementA("test2"))
}

func (suite *HashASuite) TestFindNonExistingElement() {
	assert.Equal(suite.T(), -1, suite.hashTable.findElementA("nonexistent"))

	suite.hashTable.addElementA("existing", 100)
	assert.Equal(suite.T(), -1, suite.hashTable.findElementA("nonexistent"))
}

func (suite *HashASuite) TestDeleteExistingElement() {
	suite.hashTable.addElementA("delete_me", 123)
	assert.Equal(suite.T(), 1, suite.hashTable.getSize())

	var output string
	func() {
		suite.hashTable.deleteElementA("delete_me")
	}()
	_ = output

	assert.Equal(suite.T(), 0, suite.hashTable.getSize())
	assert.Equal(suite.T(), -1, suite.hashTable.findElementA("delete_me"))
}

func (suite *HashASuite) TestDeleteNonExistingElement() {
	initialSize := suite.hashTable.getSize()
	suite.hashTable.deleteElementA("nonexistent")
	assert.Equal(suite.T(), initialSize, suite.hashTable.getSize())
}

func (suite *HashASuite) TestHashFunction() {
	hash1 := suite.hashTable.hashFunc("test", 10)
	hash2 := suite.hashTable.hashFunc("test", 10)
	assert.Equal(suite.T(), hash1, hash2)

	assert.GreaterOrEqual(suite.T(), hash1, 0)
	assert.Less(suite.T(), hash1, 10)
}

func (suite *HashASuite) TestCollisionsAndLinearProbing() {
	smallTable := NewHashA()
	smallTable.addElementA("a", 1)
	smallTable.addElementA("b", 2)
	smallTable.addElementA("c", 3)

	assert.Equal(suite.T(), 1, smallTable.findElementA("a"))
	assert.Equal(suite.T(), 2, smallTable.findElementA("b"))
	assert.Equal(suite.T(), 3, smallTable.findElementA("c"))
}

func (suite *HashASuite) TestDeleteAndReadd() {
	suite.hashTable.addElementA("temp", 999)
	assert.Equal(suite.T(), 999, suite.hashTable.findElementA("temp"))

	suite.hashTable.deleteElementA("temp")
	assert.Equal(suite.T(), -1, suite.hashTable.findElementA("temp"))

	suite.hashTable.addElementA("temp", 111)
	assert.Equal(suite.T(), 111, suite.hashTable.findElementA("temp"))
}

func (suite *HashASuite) TestPrintTable() {
	suite.hashTable.addElementA("print1", 1)
	suite.hashTable.addElementA("print2", 2)
	suite.hashTable.addElementA("print3", 3)
	suite.hashTable.deleteElementA("print1")

	assert.NotPanics(suite.T(), func() {
		suite.hashTable.printHashTableA()
	})
}

func (suite *HashASuite) TestRehash() {
	initialCap := suite.hashTable.getCapacity()
	elementsToAdd := int(float64(initialCap)*0.75) + 1
	for i := 0; i <= elementsToAdd; i++ {
		key := strings.Repeat("a", i+1)
		suite.hashTable.addElementA(key, i)
	}

	assert.Equal(suite.T(), initialCap*2, suite.hashTable.getCapacity())
}

func (suite *HashASuite) TestEdgeCases() {
	assert.Equal(suite.T(), 0, suite.hashTable.getSize())
	assert.Equal(suite.T(), -1, suite.hashTable.findElementA("anything"))

	assert.NotPanics(suite.T(), func() {
		suite.hashTable.deleteElementA("nothing")
	})
}

func (suite *HashASuite) TestInternalState() {
	suite.hashTable.addElementA("test", 123)

	cells := suite.hashTable.getCells()
	found := false
	for i, cell := range cells {
		if !cell.getIsEmpty() && !cell.getIsDeleted() && cell.getKey() == "test" {
			found = true
			assert.Equal(suite.T(), 123, cell.getData())
			assert.Equal(suite.T(), "test", cell.getKey())
			assert.False(suite.T(), cell.getIsDeleted())
			assert.False(suite.T(), cell.getIsEmpty())
			break
		}
		_ = i
	}

	assert.True(suite.T(), found, "Element should be found in the table")
}

func (suite *HashASuite) TestDeletedElementsHandling() {
	suite.hashTable.addElementA("to_delete", 456)
	suite.hashTable.deleteElementA("to_delete")

	cells := suite.hashTable.getCells()
	for _, cell := range cells {
		if cell.getKey() == "to_delete" {
			assert.True(suite.T(), cell.getIsDeleted())
			break
		}
	}

	assert.Equal(suite.T(), -1, suite.hashTable.findElementA("to_delete"))
}

File: laba3/task2/hasha_test.cpp
#include <gtest/gtest.h>
#include "structures.h"
#include "hashaddress.cpp"

using namespace std;

class HashATest : public ::testing::Test {
protected:
    void SetUp() override {
        hashTable = new HashA(5);
    }

    void TearDown() override {
        delete hashTable;
    }

    HashA* hashTable;
};

TEST_F(HashATest, AddElement) {
    hashTable->addElementA("key1", 100);
    EXPECT_EQ(hashTable->getSize(), 1);

    hashTable->addElementA("key2", 200);
    EXPECT_EQ(hashTable->getSize(), 2);
}

// Обновление значения
TEST_F(HashATest, AddElementUpdate) {
    hashTable->addElementA("key1", 100);
    hashTable->addElementA("key1", 200);

    EXPECT_EQ(hashTable->getSize(), 1);
    EXPECT_EQ(hashTable->findElementA("key1"), 200);
}

// Поиск существующих
TEST_F(HashATest, FindExistingElement) {
    hashTable->addElementA("test1", 42);
    hashTable->addElementA("test2", 84);

    EXPECT_EQ(hashTable->findElementA("test1"), 42);
    EXPECT_EQ(hashTable->findElementA("test2"), 84);
}

// Поиск несуществующих
TEST_F(HashATest, FindNonExistingElement) {
    EXPECT_EQ(hashTable->findElementA("nonexistent"), -1);

    hashTable->addElementA("existing", 100);
    EXPECT_EQ(hashTable->findElementA("nonexistent"), -1);
}

// Удаление существующих
TEST_F(HashATest, DeleteExistingElement) {
    hashTable->addElementA("delete_me", 123);
    EXPECT_EQ(hashTable->getSize(), 1);

    hashTable->deleteElementA("delete_me");
    EXPECT_EQ(hashTable->getSize(), 0);
    EXPECT_EQ(hashTable->findElementA("delete_me"), -1);
}

// Удаление несуществующих
TEST_F(HashATest, DeleteNonExistingElement) {
    testing::internal::CaptureStdout();
    hashTable->deleteElementA("nonexistent");
    const string output = testing::internal::GetCapturedStdout();
    EXPECT_TRUE(output.find("не найден") != string::npos);
}

TEST_F(HashATest, HashFunction) {
    const int hash1 = hashTable->hashFunc("test", 10);
    const int hash2 = hashTable->hashFunc("test", 10);
    EXPECT_EQ(hash1, hash2);

    EXPECT_GE(hash1, 0);
    EXPECT_LT(hash1, 10);
}

// Коллизии
TEST_F(HashATest, CollisionsAndLinearProbing) {
    HashA smallTable(3);

    smallTable.addElementA("a", 1);
    smallTable.addElementA("b", 2);
    smallTable.addElementA("c", 3);

    EXPECT_EQ(smallTable.findElementA("a"), 1);
    EXPECT_EQ(smallTable.findElementA("b"), 2);
    EXPECT_EQ(smallTable.findElementA("c"), 3);
}

TEST_F(HashATest, DeleteAndReadd) {
    hashTable->addElementA("temp", 999);
    EXPECT_EQ(hashTable->findElementA("temp"), 999);

    hashTable->deleteElementA("temp");
    EXPECT_EQ(hashTable->findElementA("temp"), -1);

    hashTable->addElementA("temp", 111);
    EXPECT_EQ(hashTable->findElementA("temp"), 111);
}

TEST_F(HashATest, PrintTable) {
    hashTable->addElementA("print1", 1);
    hashTable->addElementA("print2", 2);
    hashTable->addElementA("print3", 3);
    hashTable->deleteElementA("print1");

    testing::internal::CaptureStdout();
    hashTable->printHashTableA();
    string output = testing::internal::GetCapturedStdout();

    EXPECT_TRUE(output.find("print1") != string::npos ||
                output.find("print2") != string::npos ||
                output.find("Хэш-таблица") != string::npos);
}
// Рехэширование
TEST_F(HashATest, Rehash)
{
    const int cap = hashTable->getCapacity();
    hashTable->addElementA("hash", 1);
    hashTable->addElementA("hash2", 2);
    hashTable->addElementA("hash3", 3);
    hashTable->addElementA("hash4", 4);
    hashTable->addElementA("hash5", 5);
    EXPECT_EQ(hashTable->getCapacity(), cap * 2);
}

// Особые
TEST_F(HashATest, EdgeCases) {
    EXPECT_EQ(hashTable->getSize(), 0);

    EXPECT_EQ(hashTable->findElementA("anything"), -1);

    testing::internal::CaptureStdout();
    hashTable->deleteElementA("nothing");
    string output = testing::internal::GetCapturedStdout();
    EXPECT_TRUE(output.find("не найден") != string::npos);
}